### ISSUE TITLE
Refactor History/Visibility Archiver implementation

### ISSUE THREAD

**Issue by yycptt** [uber/cadence#3001]
Now we have three implementations for the history archiver interface (filestore, s3 and gcp) and two implementations for the visibility archiver interface (filestore and s3). They all look quite similar and share lots of common utility functions. 

We can move those common functions to the archiver package so that they can be reused. We can also consider abstract some common structure into a new component. 

Also consider adding a `context.Context` parameter to the `ValidateURI` method.

**Comment by nia-1998** [comment]
@yycptt @c-warren can I try this?

**Comment by c-warren** [comment]
Yes feel free to give this a go - I'd recommend working with @fimanishi as he is about to embark on some work that touches archival as well and he can direct you where necessary.
